File: packages/game-sdk/src/controller-bindings.ts
import type { SemanticControl } from "./types.js";

export const CONTROLLER_BUTTONS = [
  "south", "east", "west", "north", "dpad-up", "dpad-down", "dpad-left", "dpad-right",
  "left-shoulder", "right-shoulder", "left-stick", "right-stick", "start", "select",
] as const;
export const CONTROLLER_AXES = [
  "left-x", "left-y", "right-x", "right-y", "left-trigger", "right-trigger",
] as const;

export type ControllerBinding =
  | { readonly kind: "button"; readonly input: typeof CONTROLLER_BUTTONS[number]; readonly control: string }
  | { readonly kind: "axis"; readonly input: typeof CONTROLLER_AXES[number]; readonly control: string }
  | { readonly kind: "axis-button"; readonly input: typeof CONTROLLER_AXES[number]; readonly direction: -1 | 1; readonly control: string };

export interface ControllerBindings {
  readonly schema: 1;
  readonly bindings: readonly ControllerBinding[];
}

function record(value: unknown): Record<string, unknown> | undefined {
  return typeof value === "object" && value !== null && !Array.isArray(value)
    ? value as Record<string, unknown> : undefined;
}

/** Validate authored physical mappings without inventing a controller or PlayerSlot assignment. */
export function validateControllerBindings(value: unknown, controls: readonly SemanticControl[]): ControllerBindings {
  const profile = record(value);
  if (!profile || profile.schema !== 1 || Object.keys(profile).some(key => key !== "schema" && key !== "bindings") ||
    !Array.isArray(profile.bindings) || profile.bindings.length < 1 || profile.bindings.length > 64) {
    throw new TypeError("controllerBindings must have schema 1 and 1–64 bindings");
  }
  const sources = new Set<string>();
  const analogInputs = new Set<string>();
  const directionInputs = new Set<string>();
  for (const raw of profile.bindings) {
    const binding = record(raw);
    if (!binding || !["button", "axis", "axis-button"].includes(String(binding.kind))) {
      throw new TypeError("controllerBindings contains an unsupported binding kind");
    }
    const allowed = binding.kind === "axis-button" ? ["kind", "input", "control", "direction"] : ["kind", "input", "control"];
    if (Object.keys(binding).some(key => !allowed.includes(key))) throw new TypeError("controllerBindings contains unknown fields");
    const inputs: readonly string[] = binding.kind === "button" ? CONTROLLER_BUTTONS : CONTROLLER_AXES;
    if (typeof binding.input !== "string" || !inputs.includes(binding.input)) throw new TypeError("controllerBindings contains an unsupported physical input");
    if (binding.kind === "axis-button" && binding.direction !== -1 && binding.direction !== 1) {
      throw new TypeError("controllerBindings axis-button direction must be -1 or 1");
    }
    const control = controls.find(control => control.id === binding.control);
    if (!control || control.kind !== (binding.kind === "axis" ? "axis" : "button")) {
      throw new TypeError("controllerBindings must reference a declared control of the matching kind");
    }
    const key = `${binding.kind}:${binding.input}:${binding.kind === "axis-button" ? binding.direction : ""}`;
    if (sources.has(key)) throw new TypeError("controllerBindings cannot map a physical source twice");
    sources.add(key);
    if (binding.kind === "axis") analogInputs.add(binding.input);
    if (binding.kind === "axis-button") directionInputs.add(binding.input);
  }
  if ([...analogInputs].some(input => directionInputs.has(input))) {
    throw new TypeError("controllerBindings cannot map an axis as both analog and directional buttons");
  }
  return value as ControllerBindings;
}

File: apps/android/app/src/main/kotlin/dev/yougotserved/thorium/ControllerBindings.kt
package dev.yougotserved.thorium

import org.json.JSONArray
import org.json.JSONObject

/** Portable release-authored inputs; Android key/axis numbers never enter a Game Package. */
data class ControllerBinding(
    val kind: String,
    val input: String,
    val control: String,
    val direction: Int? = null,
)

data class ControllerBindings(val schema: Int = 1, val bindings: List<ControllerBinding>) {
    fun validate(controls: List<ReleaseControl>) {
        require(schema == 1 && bindings.size in 1..64) { "Invalid controller binding schema/count" }
        val sources = mutableSetOf<String>()
        bindings.forEach { binding ->
            val expectedKind = when (binding.kind) {
                "button" -> {
                    require(binding.input in BUTTONS && binding.direction == null)
                    "button"
                }
                "axis" -> {
                    require(binding.input in AXES && binding.direction == null)
                    "axis"
                }
                "axis-button" -> {
                    require(binding.input in AXES && binding.direction in setOf(-1, 1))
                    "button"
                }
                else -> error("Unknown controller binding kind")
            }
            require(controls.any { it.id == binding.control && it.kind == expectedKind }) {
                "Controller binding references an unknown control or wrong kind"
            }
            require(sources.add("${binding.kind}/${binding.input}/${binding.direction}")) {
                "Duplicate physical controller source"
            }
        }
        val analog = bindings.filter { it.kind == "axis" }.map { it.input }.toSet()
        require(bindings.none { it.kind == "axis-button" && it.input in analog }) {
            "An axis cannot be both analog and digital"
        }
    }

    fun toJson(): JSONObject = JSONObject().put("schema", schema).put(
        "bindings",
        JSONArray(bindings.map { binding ->
            JSONObject().put("kind", binding.kind).put("input", binding.input)
                .put("control", binding.control).also { value ->
                    binding.direction?.let { value.put("direction", it) }
                }
        }),
    )

    companion object {
        val BUTTONS = setOf(
            "south", "east", "west", "north", "dpad-up", "dpad-down", "dpad-left", "dpad-right",
            "left-shoulder", "right-shoulder", "left-stick", "right-stick", "start", "select",
        )
        val AXES = setOf("left-x", "left-y", "right-x", "right-y", "left-trigger", "right-trigger")

        fun parse(value: Any?, controls: List<ReleaseControl>): ControllerBindings = try {
            require(value is JSONObject && value.keys().asSequence().toSet() == setOf("schema", "bindings"))
            require(value.opt("schema") is Int)
            val entries = value.optJSONArray("bindings") ?: error("Missing controller bindings")
            require(entries.length() in 1..64)
            ControllerBindings(value.getInt("schema"), (0 until entries.length()).map { index ->
                val entry = entries.optJSONObject(index) ?: error("Invalid controller binding")
                val kind = entry.opt("kind") as? String ?: error("Missing binding kind")
                val fields = setOf("kind", "input", "control") +
                    if (kind == "axis-button") setOf("direction") else emptySet()
                require(entry.keys().asSequence().toSet() == fields)
                ControllerBinding(
                    kind, entry.opt("input") as? String ?: error("Missing input"),
                    entry.opt("control") as? String ?: error("Missing control"),
                    if (kind == "axis-button") entry.opt("direction") as? Int
                        ?: error("Invalid direction") else null,
                )
            }).also { it.validate(controls) }
        } catch (_: RuntimeException) {
            throw CatalogParseException("controllerBindings is invalid")
        }
    }
}
